packages/console/src/pages/SignInExperience/tabs/SignUpAndSignInTab/components/ConnectorSetupWarning.tsx

```diff
@@ -0,0 +1,106 @@
+import {
+  ConnectorResponse,
+  ConnectorType,
+  SignInIdentifier,
+  SignUpIdentifier,
+} from '@logto/schemas';
+import { useMemo } from 'react';
+import { useTranslation } from 'react-i18next';
+import { snakeCase } from 'snake-case';
+import useSWR from 'swr';
+
+import Alert from '@/components/Alert';
+import { RequestError } from '@/hooks/use-api';
+
+import { signUpToSignInIdentifierMapping } from '../constants';
+
+type Props = {
+  signUpIdentifier?: SignUpIdentifier;
+  signInIdentifiers?: SignInIdentifier[];
+};
+
+const getNoConnectorI18nContext = (connectorTypes: ConnectorType[]) => {
+  const { length } = connectorTypes;
+
+  if (length === 1) {
+    return connectorTypes[0]?.toLocaleLowerCase() ?? '';
+  }
+
+  if (
+    length === 2 &&
+    connectorTypes.includes(ConnectorType.Email) &&
+    connectorTypes.includes(ConnectorType.Sms)
+  ) {
+    return [ConnectorType.Email, ConnectorType.Sms].join('_or_').toLocaleLowerCase();
+  }
+
+  return '';
+};
+
+const ConnectorSetupWarning = ({ signUpIdentifier, signInIdentifiers }: Props) => {
+  const { data: connectors } = useSWR<ConnectorResponse[], RequestError>('/api/connectors');
+  const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
+
+  const requiredSignInIdentifiers = useMemo(() => {
+    if (signInIdentifiers) {
+      return signInIdentifiers;
+    }
+
+    if (signUpIdentifier) {
+      return signUpToSignInIdentifierMapping[signUpIdentifier];
+    }
+
+    return [];
+  }, [signUpIdentifier, signInIdentifiers]);
+
+  const connectorTypes = useMemo(() => {
+    return requiredSignInIdentifiers.length === 0
+      ? [ConnectorType.Social]
+      : requiredSignInIdentifiers.reduce<ConnectorType[]>((previous, current) => {
+          switch (current) {
+            case SignInIdentifier.Username: {
+              return previous;
+            }
+
+            case SignInIdentifier.Email: {
+              return [...previous, ConnectorType.Email];
+            }
+
+            case SignInIdentifier.Phone: {
+              return [...previous, ConnectorType.Sms];
+            }
+
+            default: {
+              // Note: display an TS2322 error in the IDE when the cases are not fully listed.
+              const unreachableCase: never = current;
+              throw new Error(unreachableCase);
```

Comment:
use if return could bypass the issue?

Reply:
No, because we list all cases.
Refactored with an `unreachableCaseGuardError` util to handle this logic.